### ISSUE TITLE
daemon: policy computation forcibly triggered in Cilium K8s endpoint OnUpdate function even if nothing has changed 

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#3673]
We start a new Kubernetes Informer for watching endpoint lifecycle events: https://github.com/cilium/cilium/blob/45bca6a9b6aa913c254e1790ad61af884f4db075/daemon/k8s_watcher.go#L364-L400

The documentation for `NewInformer`:
```
// NewInformer returns a Store and a controller for populating the store
// while also providing event notifications. You should only used the returned
// Store for Get/List operations; Add/Modify/Deletes will cause the event
// notifications to be faulty.
//
// Parameters:
//  * lw is list and watch functions for the source of the resource you want to
//    be informed of.
//  * objType is an object of the type that you expect to receive.
//  * resyncPeriod: if non-zero, will re-list this often (you will get OnUpdate
//    calls, even if nothing changed). Otherwise, re-list will be delayed as
//    long as possible (until the upstream source closes the watch or times out,
//    or you stop the controller).
//  * h is the object you want notifications sent to.
//
func NewInformer(
	lw ListerWatcher,
	objType runtime.Object,
	resyncPeriod time.Duration,
	h ResourceEventHandler,
)
```

Specifically: 

```
//  * resyncPeriod: if non-zero, will re-list this often (you will get OnUpdate
//    calls, even if nothing changed). Otherwise, re-list will be delayed as
//    long as possible (until the upstream source closes the watch or times out,
//    or you stop the controller).
```

Since we provide a non-zero value for resyncPeriod (5 minutes), and we have an `OnUpdate` function this means that the `OnUpdate` function is always called:

https://github.com/cilium/cilium/blob/45bca6a9b6aa913c254e1790ad61af884f4db075/daemon/k8s_watcher.go#L853-L865

The actual logic for which is encoded here: https://github.com/cilium/cilium/blob/45bca6a9b6aa913c254e1790ad61af884f4db075/daemon/k8s_watcher.go#L799

In certain cases, we forcibly trigger policy updates for the daemon:
https://github.com/cilium/cilium/blob/45bca6a9b6aa913c254e1790ad61af884f4db075/daemon/k8s_watcher.go#L841-L850

In conclusion: if nothing changed, we will call `addK8sEndpointV1`, which forcibly triggers policy updates even if nothing should change. This seems wrong to me; we should only trigger policy updates if something has changed internally within Cilium. Also, I do not think we need to call `TriggerPolicyUpdates()` with force set to 'true', because that causes all endpoints to be regenerated unconditionally.

**Comment by ianvernon** [comment]
Seems like the easiest fix for now is to change `TriggerPolicyUpdates(true)` --> `TriggerPolicyUpdates(false)` for at least the `OnUpdate` function. Not sure about `OnDelete`.

If we can compute a diff between what we have received in the past vs what we have now for endpoint updates, we can use that logic to not even call `TriggerPolicyUpdates` whatsoever.
